virtcontainers/hypervisor_plugin.go

```diff
@@ -0,0 +1,209 @@
+//
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package virtcontainers
+
+import (
+	"fmt"
+	"plugin"
+)
+
+const (
+	initFuncName                = "Init"
+	createPodFuncName           = "CreatePod"
+	startPodFuncName            = "StartPod"
+	waitPodFuncName             = "WaitPod"
+	stopPodFuncName             = "StopPod"
+	pausePodFuncName            = "PausePod"
+	resumePodFuncName           = "ResumePod"
+	addDeviceFuncName           = "AddDevice"
+	hotplugAddDeviceFuncName    = "HotplugAddDevice"
+	hotplugRemoveDeviceFuncName = "HotplugRemoveDevice"
+	getPodConsoleFuncName       = "GetPodConsole"
+	capabilitiesFuncName        = "Capabilities"
+)
+
+type pluginHypervisor struct {
+	handler    *plugin.Plugin
+	hypervisor interface{}
+}
+
+func (h *pluginHypervisor) init(pod *Pod) (err error) {
+	pluginPath := pod.config.HypervisorConfig.PluginPath
+
+	h.handler, err = plugin.Open(pluginPath)
+	if err != nil {
+		return fmt.Errorf("Failed to open plugin path %q: %v",
+			pluginPath, err)
+	}
+
+	initFunc, err := h.handler.Lookup(initFuncName)
+	if err != nil {
+		return fmt.Errorf("Failed to lookup function %q: %v",
+			initFuncName, err)
+	}
+
+	h.hypervisor, err = initFunc.(func(*Pod) (interface{}, error))(pod)
```

Comment:
Using `interface{}` is going to have a significant performance impact over using a concrete type so I think we should:

- Create a new package defining the hypervisor interface.
  ```
  package hypervisor

  type Hypervisor interface {
      CreatePod(config PodConfig) error
        : 
  }
  ```
- Import the package into the plugin and this file.

- Have each hypervisor plugin define something like the following:
  ```
  func GetHypervisorPlugin() (*hypervisor.Hypervisor, error)
  ```
- Call `h, err := plugin.Lookup("GetHypervisorPlugin")`

- Check that what you got back is actually a `hypervisor.Hypervisor` with a type assertion.

- Make calls on the returned interface:
   ```
   err := h.CreatePod(config)
   ```

That approach:

- should be a lot faster.
- will simplify the code:
  - as you only need to do the lookup once.
  - you can just call the interface methods without all the type assertions.

Reply:
I will have to try this out on a simple example first because I don't think this is going to work.
My understanding of the way this Go plugin was working was that we needed to lookup for every function or variable that we want to use from this plugin.
Basically, I think that even if you introduce this external hypervisor package, you will have to lookup for every function before you can call into them.

I'll try that and let you know.

Reply:
It works - I have example code if you're interested.

Reply:
Just tested it on my side and it works :)
This is great and I agree your solution looks way better than mine. Let me rework this PR to apply this solution !

Reply:
@jodh-intel there is one big no-go with your solution for plugin... We're gonna end up having a cycle dependency. `hypervisor_plugin.go` in virtcontainers will import the new `hypervisor` external package, but this same package will need to import virtcontainers in order to know about types like `vc.Pod`, `vc.PodConfig`, ...

Reply:
Yes, that is distressing (aka a pita). The only way round that I can think of is to sub-divide virtcontainers into smaller packages as suggested by @sameo on https://github.com/containers/virtcontainers/issues/631. That could be rather time-consuming to do at this point as it would require rework of the code and the tests.

Reply:
Yes, this is the only option (breaking virtcontainers into smaller packages) to be able to implement your solution (that I like btw). This will involve a lot of work about the code and tests, I agree, and this will likely end up with other cyclic dependency issues...

Reply:
I guess another approach would be to convert all virtcontainer types into seralised strings so that we could have:

```
package hypervisor

type PodConfig string

type Hypervisor interface {
    CreatePod(config PodConfig) error
      : 
}
```

That would result in a lot of serialising/deserialising though which won't help performance.

Reply:
I'd prefer splitting since we already want to split the api package.

Reply:
@bergwolf I think splitting is really not worth it, see my comment here: https://github.com/kata-containers/runtime/issues/66

@jodh-intel I have just realized that we actually have a way to implement what you have described without having to split anything here. We only need to export the interface `hypervisor` as `Hypervisor`. I think that as long as we are okay with having a plugin for our hypervisors, this means that we should not mind having this interface exported. In this case, this is a good reason to export one of our interface.
Following this solution, we would end up writing a dynamic library relying on virtcontainers only (no need for an intermediate external package). And we would still need an `Hypervisor` implementation `hypervisor_plugin.so` to do the glue with the actual implementation coming from the dynamic lib.
Does this make sense ??